addons/crm/models/crm_team_member.py

```diff
@@ -50,163 +72,107 @@ def _constrains_assignment_domain(self):
-                ['&', '&', ('user_id', '=', False), ('date_open', '=', False), ('team_id', '=', member.crm_team_id.id)]
-            ])
-
-            leads = self.env["crm.lead"].search(lead_domain, order='probability DESC, id', limit=lead_limit)
-
-            to_assign = member._get_assignment_quota(work_days=work_days)
-            members_data[member.id] = {
-                "team_member": member,
-                "max": member.assignment_max,
-                "to_assign": to_assign,
-                "leads": leads,
-                "assigned": self.env["crm.lead"],
-            }
-            population.append(member.id)
-            weights.append(to_assign)
-
-        leads_done_ids = set()
-        counter = 0
-        # auto-commit except in testing mode
         auto_commit = not getattr(threading.current_thread(), 'testing', False)
+        result_data = {}
         commit_bundle_size = int(self.env['ir.config_parameter'].sudo().get_param('crm.assignment.commit.bundle', 100))
-        while population and any(weights):
-            counter += 1
-            member_id = random.choices(population, weights=weights, k=1)[0]
-            member_index = population.index(member_id)
-            member_data = members_data[member_id]
-
-            lead = next((lead for lead in member_data['leads'] if lead.id not in leads_done_ids), False)
-            if lead:
-                leads_done_ids.add(lead.id)
-                members_data[member_id]["assigned"] += lead
-                weights[member_index] = weights[member_index] - 1
-
-                lead.with_context(mail_auto_subscribe_no_notify=True).convert_opportunity(
-                    lead.partner_id,
-                    user_ids=member_data['team_member'].user_id.ids
-                )
+        quota_per_member = {member: member._get_assignment_quota(force_quota=force_quota) for member in self}
+        counter = 0
+        for team in self.crm_team_id:
+            team_members = self.filtered(lambda member:
+                member.crm_team_id == team and
+                not member.assignment_optout and
+                member.assignment_max > 0 and
+                quota_per_member.get(member, 0) > 0
+            ).sorted(key=lambda m: quota_per_member.get(m, 0), reverse=True)
+            if not team_members:
+                continue
+            leads_per_member = {
+                member: self.env["crm.lead"].search(member._get_assignment_domain())
+                for member in team_members
+            }
+            result_data.update(
+                {
+                    member: {"assigned": self.env["crm.lead"], "quota": quota_per_member[member]}
+                    for member in team_members
+                }
+            )
+            leads = self.env["crm.lead"].search([
+                ('user_id', '=', False),
+                ('date_open', '=', False),
+                ('team_id', '=', team.id),
+            ], order='probability DESC, id')
```

Comment:
Hmmm.

If we're going to fetch all the leads per team, isn't it more efficient to fetch them all at once before the loop and make a dict "leads_per_team" ?

And then if we're going to do that, do we need to search in the members loop, or can we change to:

```py
leads_per_member = {
    member: leads_per_team.get(member.crm_team_id.id, self.env['crm.lead']).filtered_domain(
                   literal_eval(member.assigment_domain or '[]')
}
```

(No need to re-apply the rest of the domain as it's part of the initial domain of "leads_per_team").

Seems like it would greatly lower the necessary amount of searches.